### PR TITLE
C# bindings on .NET core 3.1

### PR DESCRIPTION
**Pull request by marat-turaev**
PR for #2272:
- All *.csproj were converted to [SDK-style](https://docs.microsoft.com/en-us/dotnet/core/tools/csproj) to [multitarget](https://docs.microsoft.com/en-us/dotnet/standard/frameworks) net452 and netcoreapp3.1
- vw_netcore.vcxproj [netcore C++/CLI](https://docs.microsoft.com/en-us/dotnet/core/porting/cpp-cli) project was added to target netcoreapp3.1
- Streamlined VS2019 migration: MSVC v142 toolset & Windows SDK 10.0 now targeted
- Other minor improvements

### CODE REVIEW

**Comment by marat-turaev** [comment]
Currently this PR is blocked by two VS bugs:

https://developercommunity.visualstudio.com/content/problem/831275/dynamic-library-project-native-lib-dependency-link.html

https://developercommunity.visualstudio.com/content/problem/900965/vs-165-preview-20-problem-with-make-heap-unchecked.html

File: vowpalwabbit/packages.config [hunk 1]
```diff
@@ -1,8 +1,8 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="boost" version="1.70.0.0" targetFramework="native" />
-  <package id="boost_program_options-vc141" version="1.70.0.0" targetFramework="native" />
-  <package id="boost_unit_test_framework-vc141" version="1.70.0.0" targetFramework="native" />
+  <package id="boost" version="1.71.0.0" targetFramework="native" />
```
Comment (lalo) [inline vowpalwabbit/packages.config]: what is prompting this update, is it required by 3.1

Comment (marat-turaev) [reply vowpalwabbit/packages.config]: It came with an upgrade to v142 MSVC toolset (VS2019): this package https://www.nuget.org/packages/boost_unit_test_framework-vc142/1.71.0 has only 1.71 version.

**Comment by marat-turaev** [comment]
Unblocking PR, as I worked around https://developercommunity.visualstudio.com/content/problem/831275/dynamic-library-project-native-lib-dependency-link.html with cb5e06c09fc7711b1cea2a62eca0e891e71943b5 and https://developercommunity.visualstudio.com/content/problem/900965/vs-165-preview-20-problem-with-make-heap-unchecked.html is non-repro on current (16.4) VS version

**Comment by thealphadollar** [comment]
@marat-turaev Any updates on this? I guess this is in pretty good shape for a re-review.

**Comment by marat-turaev** [comment]
> @marat-turaev Any updates on this? I guess this is in pretty good shape for a re-review.

Yes, this PR is open for review.

**Comment by lokitoth** [comment]
Still working on reviewing this. Going to first see if I can understand what is causing Windows CI to fail.

**Comment by marat-turaev** [comment]
> Still working on reviewing this. Going to first see if I can understand what is causing Windows CI to fail.

Due to this VS 16.5 bug: https://developercommunity.visualstudio.com/content/problem/900965/vs-165-preview-20-problem-with-make-heap-unchecked.html

It's fixed with 16.6, builds ok with preview version: https://docs.microsoft.com/en-us/visualstudio/releases/2019/release-notes-preview

**Comment by leaomatheus** [comment]
Are there plans on merging this? I'd be also interested in it to enable .NET core support for a project that takes VW as a dependency.

**Comment by smiranda** [comment]
+1 interest

**Comment by lokitoth** [comment]
Closing as completed via #3969

**Pull request closed without merging**